[CI] URLBlobContainerRetriesTests testReadBlobWithPrematureConnectionClose failing

**Build Scans:**
- [elasticsearch-periodic #4285 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/64qxditplnbtc)
- [elasticsearch-periodic #4257 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/ltqshmid5vn7a)
- [elasticsearch-periodic #4249 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/xudiztldctjco)
- [elasticsearch-periodic #4229 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/i5bp2gokcawq2)
- [elasticsearch-periodic #4225 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/fkb3lfj7swaym)
- [elasticsearch-periodic #4221 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/sacy43ey5mwkq)

**Reproduction Line:**
```
./gradlew ':modules:repository-url:test' --tests "org.elasticsearch.common.blobstore.url.URLBlobContainerRetriesTests.testReadBlobWithPrematureConnectionClose" -Dtests.seed=BA7EDBB6A38D6A6D -Dtests.locale=ia-001 -Dtests.timezone=Africa/Bissau -Druntime.java=23
```

**Applicable branches:**
7.17

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testReadBlobWithPrematureConnectionClose'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.common.blobstore.url.URLBlobContainerRetriesTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: ((a string containing "premature end of chunk coded message body: closing chunk expected" or a string containing "premature end of content-length delimited message body") or a string containing "connection closed prematurely")
     but: was "the target server failed to respond"
```

**Issue Reasons:**
- [7.17] 6 failures in test testReadBlobWithPrematureConnectionClose (1.1% fail rate in 559 executions)
- [7.17] 6 failures in step openjdk23_java-matrix (40.0% fail rate in 15 executions)
- [7.17] 6 failures in pipeline elasticsearch-periodic (35.3% fail rate in 17 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java
@@ -41,6 +41,8 @@ import java.util.concurrent.atomic.AtomicLong;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import static org.elasticsearch.test.NeverMatcher.never;
+import static org.hamcrest.Matchers.anyOf;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.either;
 import static org.hamcrest.Matchers.equalTo;
@@ -277,9 +279,12 @@ public abstract class AbstractBlobContainerRetriesTestCase extends ESTestCase {
         assertThat(exception, readTimeoutExceptionMatcher());
         assertThat(
             exception.getMessage().toLowerCase(Locale.ROOT),
-            either(containsString("read timed out")).or(containsString("premature end of chunk coded message body: closing chunk expected"))
-                .or(containsString("Read timed out"))
-                .or(containsString("unexpected end of file from server"))
+            anyOf(
+                containsString("read timed out"),
+                containsString("premature end of chunk coded message body: closing chunk expected"),
+                containsString("Read timed out"),
+                containsString("unexpected end of file from server")
+            )
         );
         assertThat(exception.getSuppressed().length, equalTo(maxRetries));
     }
@@ -308,10 +313,15 @@ public abstract class AbstractBlobContainerRetriesTestCase extends ESTestCase {
         final int maxRetries = randomInt(20);
         final BlobContainer blobContainer = createBlobContainer(maxRetries, null, null, null);
 
+        final boolean alwaysFlushBody = randomBoolean();
+
         // HTTP server sends a partial response
         final byte[] bytes = randomBlobContent(1);
         httpServer.createContext(downloadStorageEndpoint(blobContainer, "read_blob_incomplete"), exchange -> {
             sendIncompleteContent(exchange, bytes);
+            if (alwaysFlushBody) {
+                exchange.getResponseBody().flush();
+            }
             exchange.close();
         });
 
@@ -326,9 +336,14 @@ public abstract class AbstractBlobContainerRetriesTestCase extends ESTestCase {
         });
         assertThat(
             exception.getMessage().toLowerCase(Locale.ROOT),
-            either(containsString("premature end of chunk coded message body: closing chunk expected")).or(
-                containsString("premature end of content-length delimited message body")
-            ).or(containsString("connection closed prematurely"))
+            anyOf(
+                // closing the connection after sending the headers and some incomplete body might yield one of these:
+                containsString("premature end of chunk coded message body: closing chunk expected"),
+                containsString("premature end of content-length delimited message body"),
+                containsString("connection closed prematurely"),
+                // if we didn't call exchange.getResponseBody().flush() then we might not even have sent the response headers:
+                alwaysFlushBody ? never() : containsString("the target server failed to respond")
+            )
         );
         assertThat(exception.getSuppressed().length, equalTo(Math.min(10, maxRetries)));
     }

--- a/test/framework/src/main/java/org/elasticsearch/test/NeverMatcher.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/NeverMatcher.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.test;
+
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+
+public class NeverMatcher<T> extends BaseMatcher<T> {
+    @SuppressWarnings("unchecked")
+    public static <T> Matcher<T> never() {
+        return (Matcher<T>) INSTANCE;
+    }
+
+    private static final Matcher<?> INSTANCE = new NeverMatcher<>();
+
+    private NeverMatcher() {/* singleton */}
+
+    @Override
+    public boolean matches(Object actual) {
+        return false;
+    }
+
+    @Override
+    public void describeTo(Description description) {
+        description.appendText("never matches");
+    }
+}

--- a/test/framework/src/main/java/org/elasticsearch/test/NeverMatcher.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/NeverMatcher.java
@@ -1,10 +1,9 @@
 /*
  * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
- * or more contributor license agreements. Licensed under the "Elastic License
- * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
- * Public License v 1"; you may not use this file except in compliance with, at
- * your election, the "Elastic License 2.0", the "GNU Affero General Public
- * License v3.0 only", or the "Server Side Public License, v 1".
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
  */
 
 package org.elasticsearch.test;

